### PR TITLE
add channel-specific utterances, remove dispatcher allow arbitrary json output messages

### PR DESCRIPTION
**Pull request by erohmensing**
**Proposed changes**:
- removes dispatcher class from rasa_core, actions now call `send_response` on output channels directly. the handling of logging bot utterance events after action events happens via the tracker now
- adds channel-specific utterances via the `"channel"` keyword in utterance template
- add arbitrary output message formats
- rename send_custom_message() to send_elements()

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation -- still need to add more info on channel-specific utterances
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: rasa/core/actions/action.py [hunk 1]
```diff
@@ -178,10 +215,13 @@ class ActionBack(Action):
     def name(self) -> Text:
         return ACTION_BACK_NAME
 
-    def run(self, dispatcher, tracker, domain):
+    async def run(self, output_channel, nlg, tracker, domain):
```
Comment (erohmensing) [inline rasa/core/actions/action.py]: wasn't sure if this async should be added -- commenting to make sure it doesn't get lost in review.

File: .travis.yml [hunk 2]
```diff
@@ -19,7 +19,7 @@ install:
   - python -m pip install -U pip
   - pip install git+https://github.com/tmbo/MITIE.git
   - pip install -r requirements-dev.txt
-  - pip install -e .
+  - pip install -e . --no-use-pep517
```
[the remark this answers is not in the document]
Comment (tmbo) [reply .travis.yml]: already done

File: rasa/core/actions/action.py [hunk 3]
```diff
@@ -113,6 +120,24 @@ def actions_from_names(
...
+    tracker: "DialogueStateTracker", message: Dict[Text, Any]
+) -> None:
+    """Add BotUttered event to the tracker."""
+    bot_message = BotUttered(
+        text=message.get("text"),
+        data={
+            "elements": message.get("elements"),
+            "quick_replies": message.get("quick_replies"),
+            "buttons": message.get("buttons"),
+            "attachment": message.get("image") or message.get("attachment"),
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/actions/action.py]: already done

File: rasa/core/actions/action.py [hunk 4]
```diff
@@ -178,10 +215,13 @@ class ActionBack(Action):
     def name(self) -> Text:
         return ACTION_BACK_NAME
 
-    def run(self, dispatcher, tracker, domain):
+    async def run(self, output_channel, nlg, tracker, domain):  # ask Tom
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/actions/action.py]: already done :+1:

File: rasa/core/actions/action.py [hunk 5]
```diff
@@ -178,10 +215,13 @@ class ActionBack(Action):
...
         # only utter the template if it is available
-        dispatcher.utter_template("utter_back", tracker, silent_fail=True)
-        return [UserUtteranceReverted(), UserUtteranceReverted()]
+        events = await ActionUtterTemplate("utter_back", silent_fail=True).run(
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/actions/action.py]: done

File: rasa/core/actions/action.py [hunk 6]
```diff
@@ -205,12 +245,15 @@ class ActionRestart(Action):
...
         return ACTION_RESTART_NAME
 
-    async def run(self, dispatcher, tracker, domain):
+    async def run(self, output_channel, nlg, tracker, domain):
         from rasa.core.events import Restarted
 
         # only utter the template if it is available
-        await dispatcher.utter_template("utter_restart", tracker, silent_fail=True)
-        return [Restarted()]
+        events = await ActionUtterTemplate("utter_restart", silent_fail=True).run(
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/actions/action.py]: done

File: rasa/core/channels/telegram.py [hunk 7]
```diff
@@ -32,6 +33,43 @@ def __init__(self, access_token):
...
+                "prices",
+            ): "send_invoice",
+        }
+
+        for params in send_functions.keys():
+            if all(kwargs.get(p) for p in params):
+                args = [kwargs.pop(p) for p in params]
+                api_call = getattr(self, send_functions[params])
+                api_call(recipient_id, *args, **kwargs)
+                continue
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply rasa/core/channels/telegram.py]: The main reason is that `send_venue` takes [latitude, longitude, title, address], while `send_location` takes [latitude, longitude], we don't want it to send the latter if it sends the former. Will have to make sure it works correctly in this case, but i think you're right as long as they're popped from line 68

File: rasa/core/channels/telegram.py [hunk 8]
```diff
@@ -32,6 +33,43 @@ def __init__(self, access_token):
...
+                "payload",
+                "provider_token",
+                "start_parameter",
+                "currency",
+                "prices",
+            ): "send_invoice",
+        }
+
+        for params in send_functions.keys():
+            if all(kwargs.get(p) for p in params):
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply rasa/core/channels/telegram.py]: i believe `description = ""` should be valid, so i like your way 👍 can also test manually

File: rasa/core/channels/channel.py [hunk 9]
```diff
@@ -266,7 +280,10 @@ def _message(recipient_id, text=None, image=None, buttons=None, attachment=None)
             "image": image,
             "buttons": buttons,
             "attachment": attachment,
+            "custom": custom,
         }
+        if obj.get("custom"):
+            obj["custom"].setdefault("recipient_id", recipient_id)
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply rasa/core/channels/channel.py]: mm, i think i might have just missed deleting this, being leftover from something where i was originally going to replace the entire message `obj` with the custom info if there was any, and then set the recipient_id inside. I don't believe this is used anywhere

**Pull request merged**